### ISSUE TITLE
serial.end() doesn't work

### ISSUE THREAD

**Issue by aster94** [espressif/arduino-esp32#3878]
### Description:
`Serial.end()` should free the serial pins and leave the user the ability to do `digitalWrite()` on them, but it doesn't work. Instead to drive the TX pin to a LOW state i need to reset the board.

I tested the code below with the following cores:
- last stable release from platformio
- master branch
- master + #3713

I also made three logic analyzer captures: [la.zip](https://github.com/espressif/arduino-esp32/files/4441212/la.zip)
An example from one of these, as you can see it shows that the pin is driven low only once (after upload of the code or reset)

![image](https://user-images.githubusercontent.com/21184718/78613408-1f96d080-786c-11ea-885a-ed0e60593c9f.png)

### Sketch: 
```cpp

#include <Arduino.h>

HardwareSerial ser(2);
#define TX_PIN 17

void setup(){}

void loop()
{
    pinMode(TX_PIN, OUTPUT);
    delay(10);

    digitalWrite(TX_PIN, HIGH);
    delay(50);
    digitalWrite(TX_PIN, LOW);
    delay(300);
    digitalWrite(TX_PIN, HIGH);
    delay(50);

    ser.begin(9600);
    delay(10);

    for (uint8_t n = 1; n < 5; n++)
    {
        ser.write(n);
        delay(10);
    }
    ser.end();
    delay(300);
}

```

NOTE: looks like that the problem was here before #3713 and #3664

**Comment by aster94** [comment]
@lbernstone do you have any idea about what can cause this behaviour?

**Comment by lbernstone** [comment]
Turn on verbose debugging so you can see if any errors are shown.
Serial2 is instantiated by default.  Why are you not using it?
Make your delays longer- a full second whenever you attach or detach pins.
Try manually detaching after the ser.end:
`gpio_matrix_out(TX_PIN, 0x100, false, false);`

**Comment by lyusupov** [comment]
One more ticket related to this issue:
https://github.com/espressif/arduino-esp32/issues/3563#issuecomment-567199429

**Comment by aster94** [comment]
@lbernstone looks like you solved it 😁

![image](https://user-images.githubusercontent.com/21184718/79046647-88869b80-7c12-11ea-910e-45a277a38ae7.png)

At first, I tried as you suggested to put big delays, then slowly I removed all of them and no problem showed up in about 5 minutes running this sketch

I think it is an easy fix if the maintainers need I can make a PR @me-no-dev 

```cpp
#include <Arduino.h>

#define TX_PIN 17

void setup(){}

void loop()
{
    pinMode(TX_PIN, OUTPUT);

    digitalWrite(TX_PIN, HIGH);
    delay(50);
    digitalWrite(TX_PIN, LOW);
    delay(1000);
    digitalWrite(TX_PIN, HIGH);

    Serial2.begin(9600);

    for (uint8_t n = 1; n < 5; n++)
    {
        Serial2.write(n);
        delay(50);
    }
    Serial2.end();
    gpio_matrix_out(TX_PIN, 0x100, false, false);
}
```

Also, note that after the `begin` no bytes are lost
![image](https://user-images.githubusercontent.com/21184718/79046663-9805e480-7c12-11ea-81ae-8ee91f83f597.png)


Here the LA log: 
[log.zip](https://github.com/espressif/arduino-esp32/files/4464774/log.zip)

**Comment by lbernstone** [comment]
The gpio_matrix disconnect is already buried in the end method.  It just isn't working for some reason.  I'll see if I can track down where it is dropping the ball.

**Comment by marcmathys** [comment]
Saw that you are interested in a flutter app to standardize access to BT wearables - did you get anywhere.  I also find this interesting and would contribute.

**Comment by SuGlider** [comment]
Issue solved with PR #5385
